ros_bt_py/src/ros_bt_py/ros_nodes/action.py

```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
         return NodeMsg.IDLE
 
     def _feedback_cb(self, feedback) -> None:
+        self.logdebug(f"Received feedback message: {feedback}")
         with self._lock:
             self._feedback = feedback
 
     def _do_tick_wait_for_action_complete(self) -> str:
-        if self._running_goal_future is None:
-            self.logfatal("Result handle future is none! Cannot continue!")
+        if self._running_goal_handle is None or self._running_goal_future is None:
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         if self._running_goal_future.done():
             result = self._running_goal_future.result()
             if result is None:
                 self._running_goal_handle = None
                 self._running_goal_future = None
                 self._active_goal = None
+
+                self._internal_state = ActionStates.FINISHED
+
+                self.logdebug("Action result is none, action call must have failed!")
                 return NodeMsg.FAILED
-            self.outputs["result"] = result
+
+            self.outputs["result"] = result.result
             self._running_goal_handle = None
             self._running_goal_future = None
             self._active_goal = None
+
+            self._internal_state = ActionStates.FINISHED
+
+            self.logdebug("Action succeeded, publishing result!")
             return NodeMsg.SUCCEED
+
         if self._running_goal_future.cancelled():
             self._running_goal_handle = None
             self._running_goal_future = None
             self._active_goal = None
+
+            self._internal_state = ActionStates.FINISHED
+
+            self.logwarn("Action execution was cancelled by the remote server!")
             return NodeMsg.FAILED
         seconds_running = (
             self._running_goal_start_time - self.ros_node.get_clock().now()
         ).nanoseconds / 1e9
+
         if seconds_running > self.options["timeout_seconds"]:
             self.logwarn(f"Cancelling goal after {seconds_running:f} seconds!")
             self._running_goal_future.cancel()
             self._running_goal_future = None
-            status = self._do_tick_cancel_running_goal()
-            if status != NodeMsg.SUCCEED:
-                return status
+            self._internal_state = ActionStates.REQUEST_GOAL_CANCELLATION
+            return NodeMsg.SUCCEED
+
```

Comment:
We assume this cancellation does work all the time?

Reply:
Also shouldn't a timeout lead to a NodeMsg.FAILED?

Reply:
You are right, in fact we should cancel this goal asynchronously. In regards to the second point we also should return RUNNING to get into the REQUEST_GOAL_CANCELLATION state with the next tick.